File: backend/studi0trace/engines/vexel/regularity.py
"""Stage 7b: regularity across the boundary graph.

Artwork is full of lines that are meant to be parallel, perpendicular or on an
axis, and a trace that has them a fraction of a degree apart reads as sloppy
even when every line is within tolerance of its own edge. This stage looks at
every straight segment the fit produced, clusters their directions, and snaps
each cluster that carries enough length to one direction: the length-weighted
mean, made exactly horizontal or vertical when it is within the axis snap, and
made exactly perpendicular to another cluster when it is within a degree of it.

A snapped line turns about a fixed point — the node at its end if it has one,
its midpoint otherwise — and the joints it shares with neighbouring segments are
re-made: two lines meet at their new crossing, a line and a curve meet where
the old joint projects onto the new line. A line with a node at both ends
cannot turn without moving a node the other arcs were fitted to, and is left
alone. Vectorizer.AI lists exactly this under "symmetry modelling"; PolyFit
calls it regularity.
"""
from __future__ import annotations

import math

import numpy as np

from studi0trace.engines.vexel.curves import Cubic, Line, Segment, _intersect

CLUSTER_DEG = 0.5        # directions within this of each other are one cluster
CLUSTER_MIN_PX = 40.0    # a cluster snaps only when it carries this much line
PERP_DEG = 0.5           # a cluster this close to perpendicular to a bigger one is made exactly so
# px. A snap may move a line's end this far and no further: the placement knows
# an edge to a tenth of a pixel, so a line that would have to move more than
# that to join its cluster is not really parallel to it, and is left where its
# pixels are. A joint moves to the crossing with its neighbour only when that
# crossing is within the same distance of the plain projection.
END_MOVE_MAX = 0.15


def _angle(seg: Line) -> float:
    d = seg.p1 - seg.p0
    return math.degrees(math.atan2(d[1], d[0])) % 180.0


def _length(seg: Segment) -> float:
    return float(np.linalg.norm(seg.p1 - seg.p0))


def _circular_mean(angles: list[float], weights: list[float]) -> float:
    """Length-weighted mean of directions modulo 180 degrees."""
    z = sum(w * complex(math.cos(math.radians(2 * a)), math.sin(math.radians(2 * a))) for a, w in zip(angles, weights))
    if abs(z) < 1e-12:
        return angles[0]
    return (math.degrees(math.atan2(z.imag, z.real)) / 2.0) % 180.0


def _diff(a: float, b: float) -> float:
    d = abs(a - b) % 180.0
    return min(d, 180.0 - d)


def cluster_directions(lines: list[tuple[float, float]], snap_axis_deg: float) -> dict[int, float]:
    """`lines` are (angle, length). Returns, per input index, the target angle of
    its cluster, for lines whose cluster carries at least CLUSTER_MIN_PX."""
    if not lines:
        return {}
    order = sorted(range(len(lines)), key=lambda i: lines[i][0])
    clusters: list[list[int]] = [[order[0]]]
    for i in order[1:]:
        if _diff(lines[i][0], lines[clusters[-1][-1]][0]) <= CLUSTER_DEG:
            clusters[-1].append(i)
        else:
            clusters.append([i])
    # the wrap at 180: first and last clusters may be one
    if len(clusters) > 1 and _diff(lines[clusters[0][0]][0], lines[clusters[-1][-1]][0]) <= CLUSTER_DEG:
        clusters[0].extend(clusters.pop())
    targets: list[tuple[float, float, list[int]]] = []  # (angle, weight, members)
    for members in clusters:
        weight = sum(lines[i][1] for i in members)
        if weight < CLUSTER_MIN_PX:
            continue
        angle = _circular_mean([lines[i][0] for i in members], [lines[i][1] for i in members])
        if min(angle, 180.0 - angle) <= snap_axis_deg:
            angle = 0.0
        elif abs(angle - 90.0) <= snap_axis_deg:
            angle = 90.0
        targets.append((angle, weight, members))
    # perpendicular pairs: the lighter cluster takes the heavier one's angle + 90
    targets.sort(key=lambda t: -t[1])
    fixed: list[tuple[float, float, list[int]]] = []
    for angle, weight, members in targets:
        for other, _w, _m in fixed:
            if _diff(angle, (other + 90.0) % 180.0) <= PERP_DEG:
                angle = (other + 90.0) % 180.0
                break
        fixed.append((angle, weight, members))
    out: dict[int, float] = {}
    for angle, _weight, members in fixed:
        for i in members:
            out[i] = angle
    return out


def _unit(angle: float) -> np.ndarray:
    return np.array([math.cos(math.radians(angle)), math.sin(math.radians(angle))])


def _line_of(seg: Segment) -> tuple[np.ndarray, np.ndarray] | None:
    if not isinstance(seg, Line):
        return None
    d = seg.p1 - seg.p0
    n = float(np.linalg.norm(d))
    return (seg.p0, d / n) if n > 1e-9 else None


def _rejoin(line: Line, end: str, neighbour: Segment | None, new_dir: np.ndarray, anchor: np.ndarray) -> None:
    """Move `line`'s `end` onto the snapped line, and the neighbour's touching end with it."""
    old = line.p1 if end == "p1" else line.p0
    corner = anchor + new_dir * float((old - anchor) @ new_dir)
    other = _line_of(neighbour) if neighbour is not None else None
    if other is not None:
        x = _intersect(anchor, new_dir, other[0], other[1])
        if x is not None and float(np.linalg.norm(x - corner)) <= END_MOVE_MAX:
            corner = x
    delta = corner - old
    if end == "p1":
        line.p1 = corner.copy()
    else:
        line.p0 = corner.copy()
    if neighbour is not None:
        # The neighbour's touching end comes along; a curve's arm at that end
        # comes too, so the tangent it was fitted with is kept.
        if end == "p1":
            neighbour.p0 = corner.copy()
            if isinstance(neighbour, Cubic):
                neighbour.c1 = neighbour.c1 + delta
        else:
            neighbour.p1 = corner.copy()
            if isinstance(neighbour, Cubic):
                neighbour.c2 = neighbour.c2 + delta


def regularize(segment_lists: list[tuple[list[Segment], bool]], snap_axis_deg: float) -> int:
    """Snap the straight segments of every list to their cluster direction.

    Each entry is (segments, closed): an arc's fitted segments, closed when the
    arc is a loop with no node. In an open arc the first segment's start and the
    last segment's end are nodes and never move. Returns how many lines moved.
    """
    entries: list[tuple[int, int]] = []
    lines: list[tuple[float, float]] = []
    for li, (segs, _closed) in enumerate(segment_lists):
        for si, seg in enumerate(segs):
            if isinstance(seg, Line) and _length(seg) > 1e-6:
                entries.append((li, si))
                lines.append((_angle(seg), _length(seg)))
    targets = cluster_directions(lines, snap_axis_deg)
    moved = 0
    for k, (li, si) in enumerate(entries):
        if k not in targets:
            continue
        segs, closed = segment_lists[li]
        line = segs[si]
        assert isinstance(line, Line)
        if _diff(_angle(line), targets[k]) < 1e-9:
            continue
        n = len(segs)
        start_is_node = not closed and si == 0
        end_is_node = not closed and si == n - 1
        if start_is_node and end_is_node:
            continue
        prev_seg = segs[(si - 1) % n] if (closed or si > 0) else None
        next_seg = segs[(si + 1) % n] if (closed or si < n - 1) else None
        if n == 1:
            prev_seg = next_seg = None
        new_dir = _unit(targets[k])
        if float(new_dir @ (line.p1 - line.p0)) < 0.0:
            new_dir = -new_dir
        if start_is_node:
            anchor = line.p0.copy()
        elif end_is_node:
            anchor = line.p1.copy()
        else:
            anchor = 0.5 * (line.p0 + line.p1)
        # the turn must stay inside what the placement can tell: an end that
        # would have to move further than END_MOVE_MAX says this line is not
        # really in the cluster
        far = max(
            float(np.linalg.norm(anchor + new_dir * float((q - anchor) @ new_dir) - q)) for q in (line.p0, line.p1)
        )
        if far > END_MOVE_MAX:
            continue
        if not start_is_node:
            _rejoin(line, "p0", prev_seg, new_dir, anchor)
        if not end_is_node:
            _rejoin(line, "p1", next_seg, new_dir, anchor)
        moved += 1
    return moved

File: backend/vexel-rs/src/regularity.rs
//! Stage 7b: regularity across the boundary graph. Mirrors `regularity.py`.
//!
//! Every straight segment the fit produced is clustered by direction; a
//! cluster carrying enough length is snapped to one direction — its
//! length-weighted mean, exactly 0/90 within the axis snap, exactly
//! perpendicular to a heavier cluster within a degree. A snapped line turns
//! about its node end if it has one, its midpoint otherwise; the joints it
//! shares with its neighbours are re-made. A line with a node at both ends is
//! left alone.

use crate::curves::{intersect, Segment, P};

pub const CLUSTER_DEG: f64 = 0.5;
pub const CLUSTER_MIN_PX: f64 = 40.0;
pub const PERP_DEG: f64 = 0.5;
/// px; see the Python: a snap may move a line's end this far and no further.
pub const END_MOVE_MAX: f64 = 0.15;

fn sub(a: P, b: P) -> P {
    [a[0] - b[0], a[1] - b[1]]
}

fn add(a: P, b: P) -> P {
    [a[0] + b[0], a[1] + b[1]]
}

fn norm(a: P) -> f64 {
    (a[0] * a[0] + a[1] * a[1]).sqrt()
}

fn dot(a: P, b: P) -> f64 {
    a[0] * b[0] + a[1] * b[1]
}

fn angle_of(p0: P, p1: P) -> f64 {
    let d = sub(p1, p0);
    d[1].atan2(d[0]).to_degrees().rem_euclid(180.0)
}

fn diff(a: f64, b: f64) -> f64 {
    let d = (a - b).abs().rem_euclid(180.0);
    d.min(180.0 - d)
}

fn circular_mean(angles: &[f64], weights: &[f64]) -> f64 {
    let (mut re, mut im) = (0.0f64, 0.0f64);
    for (a, w) in angles.iter().zip(weights) {
        let t = (2.0 * a).to_radians();
        re += w * t.cos();
        im += w * t.sin();
    }
    if (re * re + im * im).sqrt() < 1e-12 {
        return angles[0];
    }
    (im.atan2(re).to_degrees() / 2.0).rem_euclid(180.0)
}

/// `lines` are (angle, length). Per input index, the target angle of its
/// cluster, for lines whose cluster carries at least CLUSTER_MIN_PX.
pub fn cluster_directions(lines: &[(f64, f64)], snap_axis_deg: f64) -> Vec<Option<f64>> {
    let mut out = vec![None; lines.len()];
    if lines.is_empty() {
        return out;
    }
    let mut order: Vec<usize> = (0..lines.len()).collect();
    order.sort_by(|a, b| lines[*a].0.partial_cmp(&lines[*b].0).unwrap());
    let mut clusters: Vec<Vec<usize>> = vec![vec![order[0]]];
    for &i in &order[1..] {
        let last = *clusters.last().unwrap().last().unwrap();
        if diff(lines[i].0, lines[last].0) <= CLUSTER_DEG {
            clusters.last_mut().unwrap().push(i);
        } else {
            clusters.push(vec![i]);
        }
    }
    if clusters.len() > 1 {
        let first = clusters[0][0];
        let last = *clusters.last().unwrap().last().unwrap();
        if diff(lines[first].0, lines[last].0) <= CLUSTER_DEG {
            let tail = clusters.pop().unwrap();
            clusters[0].extend(tail);
        }
    }
    let mut targets: Vec<(f64, f64, Vec<usize>)> = Vec::new();
    for members in clusters {
        let weight: f64 = members.iter().map(|&i| lines[i].1).sum();
        if weight < CLUSTER_MIN_PX {
            continue;
        }
        let angles: Vec<f64> = members.iter().map(|&i| lines[i].0).collect();
        let weights: Vec<f64> = members.iter().map(|&i| lines[i].1).collect();
        let mut angle = circular_mean(&angles, &weights);
        if angle.min(180.0 - angle) <= snap_axis_deg {
            angle = 0.0;
        } else if (angle - 90.0).abs() <= snap_axis_deg {
            angle = 90.0;
        }
        targets.push((angle, weight, members));
    }
    targets.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap());
    let mut fixed: Vec<(f64, f64, Vec<usize>)> = Vec::new();
    for (mut angle, weight, members) in targets {
        for (other, _w, _m) in &fixed {
            if diff(angle, (other + 90.0).rem_euclid(180.0)) <= PERP_DEG {
                angle = (other + 90.0).rem_euclid(180.0);
                break;
            }
        }
        fixed.push((angle, weight, members));
    }
    for (angle, _weight, members) in fixed {
        for i in members {
            out[i] = Some(angle);
        }
    }
    out
}

fn unit(angle: f64) -> P {
    [angle.to_radians().cos(), angle.to_radians().sin()]
}

fn line_of(seg: &Segment) -> Option<(P, P)> {
    if let Segment::Line { p0, p1 } = seg {
        let d = sub(*p1, *p0);
        let n = norm(d);
        if n > 1e-9 {
            return Some((*p0, [d[0] / n, d[1] / n]));
        }
    }
    None
}

/// Move `segs[si]`'s end (`p1` when `end_is_p1`) onto the snapped line, and
/// the neighbour's touching end — and a curve's arm there — with it.
fn rejoin(segs: &mut [Segment], si: usize, end_is_p1: bool, neighbour: Option<usize>, new_dir: P, anchor: P) {
    let old = match &segs[si] {
        Segment::Line { p0, p1 } => if end_is_p1 { *p1 } else { *p0 },
        _ => return,
    };
    let t = dot(sub(old, anchor), new_dir);
    let mut corner = add(anchor, [new_dir[0] * t, new_dir[1] * t]);
    if let Some(ni) = neighbour {
        if let Some((q, e)) = line_of(&segs[ni]) {
            if let Some(x) = intersect(anchor, new_dir, q, e) {
                if norm(sub(x, corner)) <= END_MOVE_MAX {
                    corner = x;
                }
            }
        }
    }
    let delta = sub(corner, old);
    if let Segment::Line { p0, p1 } = &mut segs[si] {
        if end_is_p1 {
            *p1 = corner;
        } else {
            *p0 = corner;
        }
    }
    if let Some(ni) = neighbour {
        match &mut segs[ni] {
            Segment::Line { p0, p1 } => {
                if end_is_p1 {
                    *p0 = corner;
                } else {
                    *p1 = corner;
                }
            }
            Segment::Cubic { p0, c1, c2, p1 } => {
                if end_is_p1 {
                    *p0 = corner;
                    *c1 = add(*c1, delta);
                } else {
                    *p1 = corner;
                    *c2 = add(*c2, delta);
                }
            }
            Segment::Arc { p0, p1, .. } => {
                if end_is_p1 {
                    *p0 = corner;
                } else {
                    *p1 = corner;
                }
            }
        }
    }
}

/// Snap the straight segments of every list to their cluster direction. Each
/// entry is (segments, closed); in an open list the first start and last end
/// are nodes and never move. Returns how many lines moved.
pub fn regularize(lists: &mut [(&mut Vec<Segment>, bool)], snap_axis_deg: f64) -> usize {
    let mut entries: Vec<(usize, usize)> = Vec::new();
    let mut lines: Vec<(f64, f64)> = Vec::new();
    for (li, (segs, _closed)) in lists.iter().enumerate() {
        for (si, seg) in segs.iter().enumerate() {
            if let Segment::Line { p0, p1 } = seg {
                let len = norm(sub(*p1, *p0));
                if len > 1e-6 {
                    entries.push((li, si));
                    lines.push((angle_of(*p0, *p1), len));
                }
            }
        }
    }
    let targets = cluster_directions(&lines, snap_axis_deg);
    let mut moved = 0;
    for (k, &(li, si)) in entries.iter().enumerate() {
        let Some(target) = targets[k] else { continue };
        let closed = lists[li].1;
        let segs: &mut Vec<Segment> = lists[li].0;
        let (p0, p1) = match &segs[si] {
            Segment::Line { p0, p1 } => (*p0, *p1),
            _ => continue,
        };
        if diff(angle_of(p0, p1), target) < 1e-9 {
            continue;
        }
        let n = segs.len();
        let start_is_node = !closed && si == 0;
        let end_is_node = !closed && si == n - 1;
        if start_is_node && end_is_node {
            continue;
        }
        let mut prev = if closed || si > 0 { Some((si + n - 1) % n) } else { None };
        let mut next = if closed || si < n - 1 { Some((si + 1) % n) } else { None };
        if n == 1 {
            prev = None;
            next = None;
        }
        let mut new_dir = unit(target);
        if dot(new_dir, sub(p1, p0)) < 0.0 {
            new_dir = [-new_dir[0], -new_dir[1]];
        }
        let anchor = if start_is_node {
            p0
        } else if end_is_node {
            p1
        } else {
            [0.5 * (p0[0] + p1[0]), 0.5 * (p0[1] + p1[1])]
        };
        let far = [p0, p1]
            .iter()
            .map(|q| {
                let t = dot(sub(*q, anchor), new_dir);
                norm(sub(add(anchor, [new_dir[0] * t, new_dir[1] * t]), *q))
            })
            .fold(0.0f64, f64::max);
        if far > END_MOVE_MAX {
            continue;
        }
        if !start_is_node {
            rejoin(segs, si, false, prev, new_dir, anchor);
        }
        if !end_is_node {
            rejoin(segs, si, true, next, new_dir, anchor);
        }
        moved += 1;
    }
    moved
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn clusters_snap_to_axis_and_perpendicular() {
        let t = cluster_directions(&[(0.4, 60.0), (179.95, 30.0), (89.6, 50.0), (45.3, 20.0)], 1.5);
        assert_eq!(t[0], Some(0.0));
        assert_eq!(t[1], Some(0.0));
        assert_eq!(t[2], Some(90.0));
        assert_eq!(t[3], None);
        let t = cluster_directions(&[(30.3, 100.0), (29.9, 90.0), (120.5, 45.0)], 1.5);
        assert!((t[0].unwrap() - t[1].unwrap()).abs() < 1e-9);
        assert!((t[2].unwrap() - (t[0].unwrap() + 90.0)).abs() < 1e-9);
    }

    #[test]
    fn a_line_between_two_nodes_never_turns() {
        let mut a = vec![Segment::Line { p0: [0.0, 0.0], p1: [50.0, 0.4] }];
        let mut b = vec![
            Segment::Line { p0: [0.0, 10.0], p1: [50.0, 10.0] },
            Segment::Cubic { p0: [50.0, 10.0], c1: [55.0, 12.0], c2: [58.0, 18.0], p1: [60.0, 25.0] },
        ];
        let moved = regularize(&mut [(&mut a, false), (&mut b, false)], 1.5);
        assert_eq!(moved, 0);
        assert_eq!(a[0].end(), [50.0, 0.4]);
        let mut c = vec![
            Segment::Line { p0: [0.0, 20.0], p1: [50.0, 20.1] },
            Segment::Cubic { p0: [50.0, 20.1], c1: [55.0, 22.0], c2: [58.0, 28.0], p1: [60.0, 35.0] },
        ];
        let moved = regularize(&mut [(&mut c, false), (&mut b, false)], 1.5);
        assert_eq!(moved, 1);
        assert!((c[0].end()[1] - 20.0).abs() < 1e-9);
        if let Segment::Cubic { p0, c1, .. } = c[1] {
            assert_eq!(p0, c[0].end());
            assert!((c1[1] - 21.9).abs() < 1e-9);
        } else {
            panic!("cubic expected");
        }
    }
}
